wezterm-gui/src/termwindow/render/draw.rs

```diff
@@ -14,9 +14,919 @@ impl crate::TermWindow {
+                                    width: tex_width as usize,
+                                    height: tex_height as usize,
+                                    fg_r: (linear_to_srgb(fg_r) * 255.0) as u8,
+                                    fg_g: (linear_to_srgb(fg_g) * 255.0) as u8,
+                                    fg_b: (linear_to_srgb(fg_b) * 255.0) as u8,
+                                    fg_a: (fg_a * 255.0) as u8,
+                                    bg_r: (linear_to_srgb(bg_r) * 255.0) as u8,
+                                    bg_g: (linear_to_srgb(bg_g) * 255.0) as u8,
+                                    bg_b: (linear_to_srgb(bg_b) * 255.0) as u8,
+                                    bg_a: (bg_a * 255.0) as u8,
+                                    glyph_id: tl.glyph_id,
+                                });
+                            } else {
+                                // Color emoji or background image - render with Cairo
+                                let atlas_surface = atlas.surface();
+                                ctx.save()?;
+                                ctx.translate(dest_x, dest_y);
+                                let scale_x = dest_width / tex_width;
+                                let scale_y = dest_height / tex_height;
+                                if (scale_x - 1.0).abs() > 0.001 || (scale_y - 1.0).abs() > 0.001 {
+                                    ctx.scale(scale_x, scale_y);
+                                }
+                                ctx.set_source_surface(&*atlas_surface, -tex_x1, -tex_y1)?;
+                                ctx.rectangle(0.0, 0.0, tex_width, tex_height);
+                                ctx.clip();
+                                ctx.paint()?;
+                                ctx.restore()?;
+                            }
+                        }
+                    }
+
+                    drop(vertices);
+                    drop(vertices_ref);
+                    vb_cell.next_index();
+                }
+            }
+        } // ctx dropped here
+
+        // PASS 2: Batch process all glyphs with caching
+        if !glyph_jobs.is_empty() {
+            use std::cell::RefCell;
+            use std::collections::HashMap;
+
+            // Cache key: (glyph_id, fg_color, bg_color, cell_dimensions)
+            // We cache the full cell including background, so cell size is part of the key.
+            #[derive(Hash, Eq, PartialEq, Clone, Copy)]
+            struct GlyphCacheKey {
+                glyph_id: u32,
+                fg_rgba: u32,
+                bg_rgba: u32,
+                cell_width: u16,
+                cell_height: u16,
+            }
+
+            struct CachedGlyph {
+                pixels: Vec<u8>, // Pre-rendered BGRA pixels (full cell)
+            }
+
+            thread_local! {
+                static GLYPH_CACHE: RefCell<HashMap<GlyphCacheKey, CachedGlyph>> = RefCell::new(HashMap::new());
```

Comment:
how is the cache pruned? What keeps memory usage within reasonable levels over time?